// components/CountrySelector/CountrySelector.js
/**
 * Created by roman on 19.10.17.
 */

import React from 'react';
import PropTypes from 'prop-types';

import YaGeocoder from '../../functions/ya-geocoder';
import { COUNTRIES } from '../../constants/Countries';

import './styles/bootstrap.min.css';
import './styles/flags.css';
import './styles/CountrySelector.css';

class CountrySelector extends React.Component {
    constructor(props) {
        super(props);

        this.hide = this.hide.bind(this);
        this.show = this.show.bind(this);

        let defaultCountry = '';
        if (this.props.defaultCountry) {
            defaultCountry = COUNTRIES.find(item => {
                return item.countryName === this.props.defaultCountry
            });
        }
        this.state = {
            currentCountry: defaultCountry ? defaultCountry.countryName : 'Germany',
            ISOALPHA2Code: defaultCountry ? defaultCountry.ISOALPHA2Code : 'DE',
            ISOALPHA3Code: defaultCountry ? defaultCountry.ISOALPHA3Code : 'DEU',
            ISONumericalCode: defaultCountry ? defaultCountry.ISONumericalCode : 276,
            displayedCountries: COUNTRIES,
            isListVisible: false
        };
    }
    // shouldComponentUpdate(nextProps, nextState) {
    //     if (this.state.currentCountry !== nextState.currentCountry &&
    //         this.state.ISOALPHA2Code !== nextState.ISOALPHA2Code &&
    //         this.state.ISOALPHA3Code !== nextState.ISOALPHA3Code &&
    //         this.state.ISONumericalCode !== nextState.ISONumericalCode) return true;
    // }
    componentWillUpdate(nextProps, nextState) {
        // console.log(nextState);
        this.props.getSelectedCountry(
            {
                countryName: nextState.currentCountry,
                ISOALPHA2Code: nextState.ISOALPHA2Code,
                ISOALPHA3Code: nextState.ISOALPHA3Code,
                ISONumericalCode: nextState.ISONumericalCode
            }
        );
    }
    componentDidMount() {
        if (typeof window !== 'undefined') {
            const yaGeocoder = new YaGeocoder();
            navigator.geolocation.getCurrentPosition((position) => {
                if (position) {
                    yaGeocoder.resolve(`${position.coords.longitude},${position.coords.latitude}`, (err, collection) => {
                        if (err) throw err;
                        const currentCountry = COUNTRIES.find(item => {
                            return item.ISOALPHA2Code === collection[0].country_code
                        });
                        this.setState({
                            ...this.state,
                            currentCountry: currentCountry.countryName,
                            ISOALPHA2Code: currentCountry.ISOALPHA2Code,
                            ISOALPHA3Code: currentCountry.ISOALPHA3Code,
                            ISONumericalCode: currentCountry.ISONumericalCode
                        });
                    });
                }
            });
        }
    }
    handleSelectCountry(countryName, ISOALPHA2Code, ISOALPHA3Code, ISONumericalCode) {
        this.setState({
            ...this.state,
            currentCountry: countryName,
            ISOALPHA2Code,
            ISOALPHA3Code,
            ISONumericalCode,
        });
    }
    handleChangeCountry(e) {
        const query = e.target.value.toLowerCase().trim();

        const displayedCountries = COUNTRIES.filter(function(item) {
            return item.countryName.toLowerCase().search(query) >= 0;
        });
        this.setState({
            displayedCountries
        });
    }
    show(e) {
        this.setState({
            ...this.state,
            isListVisible: true
        });
        if (typeof window !== 'undefined' && e.target.tagName.toUpperCase() !== 'INPUT') {
            document.addEventListener("click", this.hide);
        }
    }
    hide(e) {
        if (typeof window !== 'undefined' && e.target.tagName.toUpperCase() !== 'INPUT') {
            this.setState({
                ...this.state,
                isListVisible: false
            });
            document.removeEventListener("click", this.hide);
        }
    }
    render() {
        // console.log(this.state.currentCountry, this.state.ISOALPHA2Code,
        //     this.state.ISOALPHA3Code, this.state.ISONumericalCode);
        return (
            <div className="form-group">
                <div className="dropdown">
                    <div className="btn dropdown-toggle c-select" onClick={this.show}>
                        <span>
                            <img
                                className={this.state.ISOALPHA2Code ?
                                    `flag ${this.state.ISOALPHA2Code.toLowerCase()} fnone c-dropdown-flag` : 'c-dropdown-flag'}
                            />{this.state.currentCountry}
                        </span>
                    </div>
                    <div
                        className="dropdown-menu c-dropdown-menu"
                        style={this.state.isListVisible ? {display: 'block'} : {display: 'none'}}
                    >
                        <input className="c-dropdown-input" type="text" onInput={this.handleChangeCountry.bind(this)}/>
                        <div className="c-dropdown-menu-overflow">
                            <ul className="c-dropdown-ul">
                                {
                                    this.state.displayedCountries.map((item, index) => {
                                        const flag = item.ISOALPHA2Code.toLowerCase();
                                        return (
                                            <li
                                                tabIndex={index}
                                                className="dropdown-item c-dropdown-item"
                                                key={item.ISOALPHA2Code}
                                                onClick={(countryName, ISOALPHA2Code, ISOALPHA3Code, ISONumericalCode) =>
                                                    this.handleSelectCountry(
                                                        item.countryName,
                                                        item.ISOALPHA2Code,
                                                        item.ISOALPHA3Code,
                                                        item.ISONumericalCode
                                                    )
                                                }
                                            >
                                                <a data-option={`${item.ISOALPHA2Code}`}>
                                                    <img className={`flag ${flag} fnone c-dropdown-flag`} /> {item.countryName}
                                                </a>
                                            </li>
                                        )
                                    })
                                }
                            </ul>
                        </div>
                    </div>
                </div>
            </div>
        )
    }
}

CountrySelector.propTypes = {
    defaultCountry: PropTypes.string,
    getSelectedCountry: PropTypes.func
};

export default CountrySelector;
